### ISSUE TITLE
`no_route` error causes lnd to cache payment details and always fail

### ISSUE THREAD

**Issue by luizParreira** [lightningnetwork/lnd#5029]
### Background

Some random issue caused my node to disconnect from one of the peers we were connected to (ACINQ). Once we disconnected, all channels we have opened with them became **inactive** and as soon as we tried to send any amount of sats to a Phoenix Wallet, it would give us a `no_route` error.

### Your environment

* `0.12.0-beta`
* `Linux lnd-0 4.19.112+ #1 SMP Sat Oct 10 13:45:37 PDT 2020 x86_64 Linux`
* `0.20.1`

### Steps to reproduce

1. Have open funded channels with a specific node
2. If the connection to this node fail for any random reason
3. Try and pay an invoice generated by the other node
4. You should get `no_route` error?
5. Reconnect to the other node (peer)
6. Make sure channels are active and funded
7. Try and pay the same invoice
8. You should still get a `no_route` error

### Expected behaviour

Why do I still get `no_route` error even after reconnecting to the peer and having the channels active? Shouldn't LND know that the channel has been reactivated or maybe even retry to make the payment, instead of saving some state about an error? From what I understood, if you save the state of the error and always assume this will fail, you are assuming nodes will never be able to recover and if one fails, it is dead forever. Is that an expected behaviour? 

### Actual behaviour

`LND` persists some state about the payment that failed, and even though I reconnect to the node and retry making the payment multiple times. It always fails. It only works after running the `lncli resetmc` command.

**Comment by 3nprob** [comment]
I'm getting this to. `queryroutes` find a path (taking a couple of seconds to do so), attempting the payment fails instantaneousl, seemingly without even looking at the graph. I *know* there is a sufficient connected  liquiditypath between the nodes.

**Comment by luizParreira** [comment]
@3nprob I may have left out a small part of what I did: In order to reconnect to the peer that was disconnected, I had to restart `lnd`. When I tried to manually disconnect and then reconnect to the peer, the reconnection would never work. Once I restarted `lnd`, it was able to reconnect to the peer and when I ran `resetmc`, it started working again.

**Comment by 3nprob** [comment]
Hmm, I tried that and other things too, no success yet :/

**Comment by Roasbeef** [comment]
> . Once we disconnected, all channels we have opened with them became inactive and as soon as we tried to send any amount of sats to a Phoenix Wallet, it would give us a no_route error.

As you've noted, you can reset the mission control state to wipe things from scratch. Overtime the weights will decay meaning it'll eventually re-try those routes. There's an _existing_ issue w/ the way the Phoenix wallet handles their "just in time" channel opening which at times can cause unnecessary failures, that's being tracked here #2066. 

If the channels are _inactive_, then that means the other side never finished the handshake required to enable the channels again. We apply failures differently when they occur at the outbound channels compared to if they happen elsewhere in the network.

**Comment by 3nprob** [comment]
Thanks for explaning @Roasbeef  - the channels ar from all I can tell open, active, funded in the right direction, and I have retried for days.

**Comment by 3nprob** [comment]
OK, after a lot of trial and error we were actually able to send a handful of payments by manually looking up paths on 1ml.com, putting the node ids manually in `buildroutes | sendtoroute`. No amount of `payinvoice` or `queryroutes | sendtoroute` with the same amounts seem able to find the same paths, even when specifying `--outgoing_chan_id` and `--last_hop` with the same ones.

not using `--use_mc` in `queryroutes` will outout paths with channels that don't work (one-sided capacity in the wrong direction), wile using `--use_mc` always fails instantenaously (as in, immediatly, zero delay)

**Comment by 3nprob** [comment]
Have done multiple attempts through various methods between separate peers pvver the past week. Occasionally, a payment *does* go through (though always requiring a manuallly specified route - `sendpayment`/`payinvoice` always failes instantaneously, even if done right before a later successful payment with the same details, but a manually built route).

It also seems that if I in the first failed attempt manually specify `--outgoing_chan(_)id` / `--last_hop`, lnd will then permanently mark the payment hash as failed,refusing to even attempt to look at alyternative routes for future attempts.

Example log (>12h since last payment attempt):

```
22:01:48.316 [DBG] CRTR: Payment deadbeef in state terminate=false, active_shards=0, rem_value=1000000000 mSAT, fee_limit=1000000000 mSAT
22:01:48.317 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=1000000000 mSAT
22:01:48.317 [DBG] CRTR: Pathfinding absolute attempt cost: 1100 sats
22:01:48.318 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=891.943µs
22:01:48.318 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=500000000 mSAT
22:01:48.319 [DBG] CRTR: Pathfinding absolute attempt cost: 600 sats
22:01:48.319 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=877.636µs
22:01:48.319 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=250000000 mSAT
22:01:48.320 [DBG] CRTR: Pathfinding absolute attempt cost: 350 sats
22:01:48.320 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=777.949µs
22:01:48.321 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=125000000 mSAT
22:01:48.321 [DBG] CRTR: Pathfinding absolute attempt cost: 225 sats
22:01:48.322 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=875.813µs
22:01:48.322 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=62500000 mSAT
22:01:48.323 [DBG] CRTR: Pathfinding absolute attempt cost: 162.5 sats
22:01:48.323 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=815.991µs
22:01:48.323 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=31250000 mSAT
22:01:48.324 [DBG] CRTR: Pathfinding absolute attempt cost: 131.25 sats
22:01:48.324 [DBG] CRTR: Pathfinding perf metrics: nodes=1, edges=1, time=810.379µs
22:01:48.324 [DBG] CRTR: PaymentSession(deadbeef): pathfinding for amt=15625000 mSAT
22:01:48.325 [DBG] CRTR: Pathfinding absolute attempt cost: 115.625 sats
22:01:48.352 [DBG] CRTR: Pathfinding perf metrics: nodes=3, edges=725, time=26.985323ms
22:01:48.352 [DBG] CRTR: PaymentSession(deadbeef): not splitting because minimum shard amount 10000000 mSAT has been reached
22:01:48.352 [WRN] CRTR: Failed to find route for payment deadbeef: unable to find a path to destination
22:01:48.352 [DBG] CRTR: Marking payment deadbeef permanently failed with no route: no_route
22:01:48.362 [DBG] CRTR: Payment deadbeef in state terminate=true, active_shards=0, rem_value=1000000000 mSAT, fee_limit=1000000000 mSAT
22:01:48.362 [ERR] CRTR: Payment with hash deadbeef failed: no_route
```

In this scenario, the sending node has around a dozen outgoing channels, most of which have a path leading to one of the handful of channels with sufficient inbound capacity on the receiving node in some way.

**Comment by 3nprob** [comment]
Decreased the configuration value for `routerrpc.minrtprob` and attempts are now being made.

**Comment by sangaman** [comment]
I just ran into this issue where I was getting "no route" errors on a simple 2 hop A->B->C payment where all channels involved were active and balanced. `queryroutes` was returning nothing and paying invoices was failing. I suspect the cause was due to a period of downtime where a number of payments using that route failed, possibly causing lnd to ignore that route.

I resolved it by calling `queryroutes` with the `--use_mc` flag, that returned the route i expected and all of a sudden paying invoices started working again.
